### PR TITLE
React Query for Object Storage Bug Fixes

### PR DESCRIPTION
**Pull request by bnussman**
## Description

- Fixes deletion not showing correct object after a delete (Found by @HanaXu)
- Fixes multiple uploads not showing in list (Found by @DevDW)

## How to test

- Test the two fixes
  - Upload many objects at once and sure they all show in the table
  - Upload two objects, then delete the first. One object should remain

### CODE REVIEW

**Review by HanaXu — COMMENTED** [review]
When replacing objects (file/folder), the object gets added to the table until you refresh

This branch:

https://user-images.githubusercontent.com/14323019/155023510-54005b5f-c971-47b5-ab57-c4cf05a0f3e1.mov

Prod:

https://user-images.githubusercontent.com/14323019/155023560-d6bc8baa-f483-4526-854e-91ffec51fa0b.mov

**Review by DevDW — COMMENTED** [review]
An error pops up in the console on every object add or delete:

Screen Shot 2022-02-21 at 4 23 17 PM

**Review by HanaXu — COMMENTED** [review]
The object size/last modified isn't updating when it's replaced

https://user-images.githubusercontent.com/14323019/155028832-2f7efe06-25ca-46be-94b8-51e27f39f06d.mov

Prod:

https://user-images.githubusercontent.com/14323019/155028863-a754b645-0369-4114-82c0-46cdd03b48f1.mov

**Review by bcmertz — COMMENTED** [review]
Getting 403 on the PUT request to replace a file, with XHR throwing CORS errors "blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource."

**Comment by bnussman** [comment]
@bcmertz what is the name of the file that caused the 403?

**Comment by bcmertz** [comment]
@bnussman seldon (1).png, it's not listed as a file in Object Storage, it was previously uploaded then deleted. Perhaps not on our end?

**Comment by bnussman** [comment]
Files like that (`splash (2).png` for example) have had problems already in production. This bug is out of the scope of this ticket.

**Review by HanaXu — COMMENTED** [review]
Adding/deleting objects in a bucket makes a GET request for every bucket. Prod behavior is to only make a GET request for the updated bucket

**Comment by bcmertz** [comment]
@bnussman One small issue about updating number of objects in bucket after deleting

1. upload object
2. delete object
3. return to bucket landing
4. see number of objects not updated to reflect removed object

Requires refresh to update

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx [hunk 1]
```diff
@@ -224,11 +239,33 @@ export const BucketDetail: React.FC = () => {
...
+
+    const copy = [...data.pages];
+
+    const dataCopy = [...copy[copy.length - 1].data];
+
+    dataCopy.push(object);
+
+    copy[copy.length - 1].data = dataCopy;
+
+    updateStore(copy);
```
Comment (codyfinn) [inline packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:268]:
```
const pagesCopy = data.pages.map((page) => {
  const foundObjectIndex = page.data.findIndex(
    (_object) => _object.name === object.name
  );
  if (foundObjectIndex > -1) {
    const pageCopy = { ...page };
    pageCopy[foundObjectIndex] = object;
    return pageCopy;
  }
  return page;
  });

updateStore(pagesCopy);
```

Comment (codyfinn) [reply packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:268]: Would something like this work?

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx [hunk 2]
```diff
@@ -141,6 +144,13 @@ export const BucketDetail: React.FC = () => {
     setObjectDetailDrawerOpen(true);
   };
 
+  // If a user deletes many objects in a short amount of time,
+  // we don't want to fetch for every delete action. Refetch bucket
+  // intially but debounce on 5 seconds.
+  const debouncedUpdateBucket = debounce(5000, true, () =>
+    updateBucket(clusterId, bucketName)
+  );
```
Comment (bnussman) [inline packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:152]:
```suggestion
  const debouncedUpdateBucket = debounce(3000, false, () =>
    updateBucket(clusterId, bucketName)
  );
```

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx [hunk 3]
```diff
@@ -141,6 +144,13 @@ export const BucketDetail: React.FC = () => {
     setObjectDetailDrawerOpen(true);
   };
 
+  // If a user deletes many objects in a short amount of time,
+  // we don't want to fetch for every delete action. Refetch bucket
+  // intially but debounce on 5 seconds.
```
Comment (DevDW) [inline packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:149]: `initially`

"...debounce on 3 seconds."

**Review by DevDW — APPROVED** [review]

**Review by bcmertz — APPROVED** [review]
LGTM!

**Review by HanaXu — APPROVED** [review]
LGTM! 🚢

**Review by DevDW — APPROVED** [review]

**Pull request merged**